sql/core/src/main/scala/org/apache/spark/sql/ui/SQLListener.scala

```diff
@@ -0,0 +1,340 @@
+
+  // VisibleForTesting
+  val executionIdToData = mutable.HashMap[Long, SQLExecutionUIData]()
+
+  /**
+   * Maintain the relation between job id and execution id so that we can get the execution id in
+   * the "onJobEnd" method.
+   */
+  private val jobIdToExecutionId = mutable.HashMap[Long, Long]()
+
+  private val stageIdToStageMetrics = mutable.HashMap[Long, SQLStageMetrics]()
+
+  private val failedExecutions = mutable.ListBuffer[SQLExecutionUIData]()
+
+  private val completedExecutions = mutable.ListBuffer[SQLExecutionUIData]()
+
+  // VisibleForTesting
+  def executionIdToDataSize: Int = synchronized {
+    executionIdToData.size
+  }
+
+  // VisibleForTesting
+  def jobIdToExecutionIdSize: Int = synchronized {
+    jobIdToExecutionId.size
+  }
+
+  // VisibleForTesting
+  def stageIdToStageMetricsSize: Int = synchronized {
+    stageIdToStageMetrics.size
+  }
+
+  private def trimExecutionsIfNecessary(
+      executions: mutable.ListBuffer[SQLExecutionUIData]): Unit = {
+    if (executions.size > retainedExecutions) {
+      val toRemove = math.max(retainedExecutions / 10, 1)
+      executions.take(toRemove).foreach { execution =>
+        for (executionUIData <- executionIdToData.remove(execution.executionId)) {
+          for (jobId <- executionUIData.jobs.keys) {
+            jobIdToExecutionId.remove(jobId)
+          }
+          for (stageId <- executionUIData.stages) {
+            stageIdToStageMetrics.remove(stageId)
+          }
+        }
+      }
+      executions.trimStart(toRemove)
+    }
+  }
+
+  override def onJobStart(jobStart: SparkListenerJobStart): Unit = {
+    val executionId = jobStart.properties.getProperty(SQLExecution.EXECUTION_ID_KEY)
+    if (executionId == null) {
+      // This is not a job created by SQL
+      return
+    }
+    val jobId = jobStart.jobId
+    val stageIds = jobStart.stageIds
+
+    synchronized {
+      activeExecutions.get(executionId.toLong).foreach { executionUIData =>
+        executionUIData.jobs(jobId) = JobExecutionStatus.RUNNING
+        executionUIData.stages ++= stageIds
+        // attemptId must be 0. Right?
+        stageIds.foreach(stageId =>
+          stageIdToStageMetrics(stageId) = SQLStageMetrics(stageAttemptId = 0))
+        jobIdToExecutionId(jobId) = executionUIData.executionId
+      }
+    }
+  }
+
+  override def onJobEnd(jobEnd: SparkListenerJobEnd): Unit = synchronized {
+    val jobId = jobEnd.jobId
+    for (executionId <- jobIdToExecutionId.get(jobId);
+         executionUIData <- executionIdToData.get(executionId)) {
+      jobEnd.jobResult match {
+        case JobSucceeded => executionUIData.jobs(jobId) = JobExecutionStatus.SUCCEEDED
+        case JobFailed(_) => executionUIData.jobs(jobId) = JobExecutionStatus.FAILED
+      }
+      if (executionUIData.completionTime.nonEmpty && !executionUIData.hasRunningJobs) {
+        // onExecutionEnd happens before this onJobEnd and we are the last job, so we should update
+        // the execution lists.
```

Comment:
I would just say
```
// We are the last job of this execution, so mark the execution as finished. Note that
// `onExecutionEnd` also does this, but currently that can be called before `onJobEnd`
// since these are called on different threads.
```